Make MapHeaderMap completely case insensitive

grandbora opened the pull request:
Http header field names are case-insensitive ([rfc2616#4.2](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2)). Therefore [MapHeaderMap](https://github.com/twitter/finagle/blob/develop/finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala#L60) is designed to be case-insensitive. However only a subset of methods are obeying to this contract.

For example with the current implementation the tests below fail;

```scala
  test("test set and getAll") {
    val map = MapHeaderMap()
    map.set("KEY", "1")
    assert(map.getAll("key") == List("1"))
  }

  test("test +=") {
    val map = MapHeaderMap()
    map += ("KEY" -> "1")
    assert(map.getAll("key") == List("1"))
  }
```
```bash
[info] - test set and getAll *** FAILED ***
[info]   List() did not equal List("1") (MapHeaderMapTest.scala:13)
[info] - test += *** FAILED ***
[info]   List() did not equal List("1") (MapHeaderMapTest.scala:19)

```

There are easier solutions for this problem (details are below), however in order to be backward compatible (to a sufficient degree) this pr introduces a slightly more complicated logic.

In this pr, all the headers are stored with their original header names, but they are grouped by the [canonical name of the header](https://github.com/grandbora/finagle/blob/make-headermap-case-insensitive/finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala#L65) on write (see `set`, `add` and `+=` methods). 

In order to keep things sane, I created the [HeaderValuePair](https://github.com/grandbora/finagle/blob/make-headermap-case-insensitive/finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala#L60) abstraction, which is an implementation detail and is not visible to library users.

---- 

For the curious, other possible solutions;

* **Modify on write**
On write, lowercase every header field name (set, add, +=, apply). In [MapHeaderMap#apply](https://github.com/twitter/finagle/blob/develop/finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala#L113) method this approach has been taken but it is not propagated to other write methods.

* **Merge on read**
 On write, use given header field names as they are. On read, find every entry that matches the given header field name an a case-insensitive manner and merge their values.

* **Use a case-insensitive data structure as underlying storage object**
Delegate the responsibility of handling case-insensitivity to the underlying object. Use something like `new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);`


The disadvantage of first approach is that it relies on `.toLowerCase`. In theory [header field names can contain any ISO-8859-1 character](https://tools.ietf.org/html/rfc5987#section-1). I am not sure if there would be some edge cases in the behavior of `.toLowerCase` if a non ASCII character appears in the header field name (not even sure if this is a valid concern).

Second approach doesn't have the disadvantage mentioned above, as long as it uses `String#equalsIgnoreCase` method. However keeping various representations in the underlying storage, for a single header seems like an unnecessary complication.

Third approach can actually be the cleanest implementation, however I don't know if using case-insensitive TreeMap has performance impacts or if it is feasible.

@fwbrasil care to take a look :)

kevinoliver commented:
I haven't read the diff yet, but one consideration is how this changes iteration of existing keys. Specifically, if you had a user that did `Map.add("UP", "X")` and then iterated over the keys, they'd now see "up" instead of "UP". 

I _think_ that could in turn break users who need to look at all the headers to calculate a signature, for example in OAuth 1.1.

grandbora commented:
@kevinoliver thanks for looking into this. You pointed a valid concern. What do you think if I implement the second approach (Merge on read)? Would that be ok?

kevinoliver commented:
I can't think of a way to do merge on read, but maybe you have ideas.

grandbora commented:
Hey @kevinoliver , I gave it a shot and introduced a slightly more complex logic to work around the problem you described. I updated the pr description and added some tests for backward compatibility. I'd appreciate if you can have a look.

kevinoliver commented on finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala:
```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
+}
+
+private[finagle] object HeaderValuePair {
+  def canonicalName(header: String) = header.toLowerCase
```
the default encoding will be the platform if unspecified, so we should use `c.t.io.Charsets.UsAscii` i believe.

grandbora replied:
I was aware of that point, I saw no harm in leaving it to the default because I think as long as it is consistent through out the life cycle of the running app (or most likely within the life cycle of a single request) it *should* be fine. So that was my reasoning.

However I am no expert on this stuff, so tried to add what you suggested. It didn't work out because it expects a `Locale` not a `Charset`. I ended up using `Locale.US`. If there is a twitter convention for this please let me know.

kevinoliver commented on finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala:
```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
-class MapHeaderMap(underlying: mutable.Map[String, Seq[String]]) extends HeaderMap {
+class MapHeaderMap extends HeaderMap {
+
+  private val underlying = mutable.Map.empty[String, List[HeaderValuePair]]
```
nitpick: use `private[this]`

kevinoliver commented on finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala:
```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
+  // This is here to provide backward compatibility
+  def this(underlying: mutable.Map[String, Seq[String]]) {
+    this
+    for ((k, vs) <- underlying; v <- vs) yield{
```
style nit: add a space after `yield`

kevinoliver commented on finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala:
```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
     this
   }
 
   override def keys: Iterable[String] =
```
do you think we should document that these will return the keys in a normalized form (here and below)?

grandbora replied:
Correct me if I am missing something here. The purpose of the rework was to make sure that we return the original header names, not the normalized ones. [There is a test case specifically for that](https://github.com/grandbora/finagle/blob/make-headermap-case-insensitive/finagle-http/src/test/scala/com/twitter/finagle/http/MapHeaderMapTest.scala#L113-L124). So behavior here should not change.

kevinoliver replied:
yeah, you're right.

kevinoliver commented on finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala:
```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
+  private val underlying = mutable.Map.empty[String, List[HeaderValuePair]]
+
+  // This is here to provide backward compatibility
+  def this(underlying: mutable.Map[String, Seq[String]]) {
```
hmm i realize why you left it, but i think its a bit deceptive to the end-user. given that they pass in a mutable map, there is likely some expectation that their modifications to the map are reflected in this `HeaderMap`. but they won't.

this is a big enough change in behavior that i think we should at a minimum mark this as deprecated, and strongly consider breaking the API by removing the constructor.

grandbora replied:
You have a valid point here. I was going to suggest adding a deprecation here. But I personally prefer breaking the api by removing this constructor. Reasons:
* there is already an apply method which takes tuples. I would expect that most people are using that over this constructor.
* this is a pretty easy fix on the users' side. I don't think it worths the trouble of deprecating it.

Let me know which way you want to go with. Happy to do the changes.

kevinoliver commented:
@grandbora this looks like a good approach. i think its worthy of a note in `CHANGES` under the "Runtime Behavior Changes" section, if you could add that as well. or you could call it a "Bug Fix" — either way is fine.

grandbora commented:
I updated the `CHANGES` accordingly.

kevinoliver commented on finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala:
```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
+
+  private[this] val underlying = mutable.Map.empty[String, List[HeaderValuePair]]
+
+  // This is here to provide backward compatibility
```
please either deprecate or remove. as mentioned, i think this API is too weird now.

grandbora replied:
Removed https://github.com/twitter/finagle/pull/473/commits/8f01e6bf3b02809cc60b242ea9d6c1aea1640d26

kevinoliver commented:
thanks @grandbora. working on pulling this in locally. looks like there is some usage of the mutable map constructor so it may take some time.

kevinoliver commented:
This just landed in `develop` with commit https://github.com/twitter/finagle/commit/603c0226029f8cc9da2ea628e793e80a91c170ef

Thanks for the contribution @grandbora.

grandbora commented:
Yay! Looking forward to use this on prod. Thanks @kevinoliver
